### ISSUE TITLE
Logging in the compiler

### ISSUE THREAD

**Issue by philberty** [Rust-GCC/gccrs#248]
We currently use a lot of printf but this is messy it would be ideal to have some kind of logging framework.

There are loggers in the jit framework and the analyzer framework, @davidmalcolm @tschwinge do you guys have any opinions on logging in GCC?  Is there room to create some kind of generic logger for components in GCC? Its something I could spend some time on if it makes sense using the analyzer/jit ones as the base.

**Comment by davidmalcolm** [comment]
FWIW the analyzer logger was a copy-and-paste of the one in the jit subdirectory.

I'd be in favor of unifying them; I think I kept them separate to make analyzer patch review easier (IIRC).

I use the analyzer logging many times a day when debugging; it's very useful to be able to send the logs to stderr when stepping through the code in the debugger.

**Comment by philberty** [comment]
Hi David, I have created extracted the logger out here in this commit: https://github.com/Rust-GCC/gccrs/commit/35d763f6c36d94594282f9d8bef938e71a3e8b28

It would be nice to get a review. I just want to test it out a little more tomorrow and was going to post it to gcc-patches.

What do you think? @davidmalcolm and @tschwinge

**Comment by davidmalcolm** [comment]
On Wed, 2021-03-03 at 12:33 -0800, Philip Herron wrote:
> Hi David, I have created extracted the logger out here in this
> commit: 
> https://github.com/Rust-GCC/gccrs/commit/35d763f6c36d94594282f9d8bef938e71a3e8b28
> 
> It would be nice to get a review. I just want to test it out a little
> more tomorrow and was going to post it to gcc-patches.
> 
> What do you think? @davidmalcolm and @tschwinge 

Mostly looks promising (for gcc 12), but...

--- a/gcc/jit/libgccjit.c
+++ b/gcc/jit/libgccjit.c
@@ -25,6 +25,7 @@ along with GCC; see the file COPYING3.  If not see
 #include "typed-splay-tree.h"
 #include "cppbuiltin.h"
 #include <pthread.h>
+#include "diagnostic.h"
 
 #include "libgccjit.h"
 #include "jit-recording.h"
@@ -2891,9 +2892,9 @@ gcc_jit_context_set_logfile (gcc_jit_context *ctxt,
   RETURN_IF_FAIL ((flags == 0), ctxt, NULL, "flags must be 0 for now");
   RETURN_IF_FAIL ((verbosity == 0), ctxt, NULL, "verbosity must be 0 for now");
 
-  gcc::jit::logger *logger;
+  gcc::logger *logger;
   if (logfile)
-    logger = new gcc::jit::logger (logfile, flags, verbosity);
+    logger = new gcc::logger (logfile, flags, verbosity, *global_dc->printer, "JIT");
   else
     logger = NULL;
   ctxt->set_logger (logger);

I see that libgccjit.c in gcc_jit_context_set_logfile you add a use of
*global_dc->printer, but this happens outside of a mutex, and the data
itself may well not be initialized yet; gcc_jit_context_set_logfile is
typically called before toplev_main.

What does this get used for in the merged implementation?

Dave

**Comment by philberty** [comment]
> On Wed, 2021-03-03 at 12:33 -0800, Philip Herron wrote: Hi David, I have created extracted the logger out here in this commit: [35d763f](https://github.com/Rust-GCC/gccrs/commit/35d763f6c36d94594282f9d8bef938e71a3e8b28) It would be nice to get a review. I just want to test it out a little more tomorrow and was going to post it to gcc-patches. What do you think? @davidmalcolm and @tschwinge
> Mostly looks promising (for gcc 12), but...
> […](#)
> --- a/gcc/jit/libgccjit.c +++ b/gcc/jit/libgccjit.c @@ -25,6 +25,7 @@ along with GCC; see the file COPYING3. If not see #include "typed-splay-tree.h" #include "cppbuiltin.h" #include <pthread.h> +#include "diagnostic.h" #include "libgccjit.h" #include "jit-recording.h" @@ -2891,9 +2892,9 @@ gcc_jit_context_set_logfile (gcc_jit_context *ctxt, RETURN_IF_FAIL ((flags == 0), ctxt, NULL, "flags must be 0 for now"); RETURN_IF_FAIL ((verbosity == 0), ctxt, NULL, "verbosity must be 0 for now"); - gcc::jit::logger *logger; + gcc::logger *logger; if (logfile) - logger = new gcc::jit::logger (logfile, flags, verbosity); + logger = new gcc::logger (logfile, flags, verbosity, *global_dc->printer, "JIT"); else logger = NULL; ctxt->set_logger (logger); I see that libgccjit.c in gcc_jit_context_set_logfile you add a use of *global_dc->printer, but this happens outside of a mutex, and the data itself may well not be initialized yet; gcc_jit_context_set_logfile is typically called before toplev_main. What does this get used for in the merged implementation? Dave

When I was extracting the code out, I was trying to preserve the behaviour in the jit-logger here: https://github.com/Rust-GCC/gccrs/blob/e76855ab401d70361cc8c1738d95127b6f45df86/gcc/jit/jit-logging.c#L46 where it calls print_version with "JIT:" so I added another argument to the constructor for this and added it to the analyzer one too. The analyzer logger called print version with an empty string we could just default to this if you prefer?

**Comment by davidmalcolm** [comment]
I see that you added the string argument to logger's ctor.  I think it's only ever used in the ctor, so I don't think it needs to be a field.  Usage of `std::string` is controversial within the GCC codebase so you might want to convert it to just a `const char *`.

What I was more concerned about was the `*global_dc->printer` argument - is it because analyzer's logger has `m_pp`, whereas the jit logger doesn't?  The analyzer logging sometimes makes use of `m_pp`, but the jit logging doesn't, so maybe we could make it valid to pass in `NULL` for the `pretty_printer`, making this optional (and make this be a pointer rather than a reference?).

Have you tested the jit code with this patch?  The only testcase that uses `gcc_jit_context_set_logfile` seems to be `test-nested-contexts.c`.   See https://gcc.gnu.org/onlinedocs/jit/internals/index.html#running-the-test-suite for notes on running just one jit testcase.

**Comment by philberty** [comment]
Yesterday I did some testing with -fanalyzer, to then realise the jit test suite was failing. I have pushed this commit https://github.com/Rust-GCC/gccrs/commit/6b4ab8a444a9d925871ef09fa51fad9f34f1eb5f which fixes those failures.

I created a second constructor within the logger that matches the original jit logger and changed the pretty printer to be a pointer like you say so it can be null. There were some tweaks to check if we had a pretty printer or not in the case for log_va_partial and end_log_line.

The output for analyzer looks the same as before and the jit test suite is passing are there any other tests you think I should try before posting to gcc-patches?

**Comment by davidmalcolm** [comment]
Thanks.  If I'm reading the patch correctly I think there's a change of behavior: previously all jit logging lines were prefixed with `JIT: `, whereas I think that's been dropped in the patched version.  The prefixes are very useful when debugging the jit when it's linked into a big existing project, so you can quickly see what lines are coming from the jit logging versus the existing project's logging.  The analyzer doesn't bother with a prefix because the messages are grouped together in one big clump, whereas the jit log messages are more likely to be intermingled with messages from the thing libgccjit is linked into.

One other suggestion, for when you take this to the gcc-patches: the commit message is rather vague about why you want to do this.  I think you'll get a better reception if you reword it to spell out that gccrs wants to use this (and, ideally, why).

**Comment by davidmalcolm** [comment]
For reference, our discussion about this on gcc-patches is: https://gcc.gnu.org/pipermail/gcc-patches/2021-March/566295.html

**Comment by philberty** [comment]
2nd version of patch was available on: https://gcc.gnu.org/pipermail/gcc-patches/2021-March/566432.html
